// @wordpress/rich-text/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "applyFormat", {
  enumerable: true,
  get: function get() {
    return _applyFormat.applyFormat;
  }
});
Object.defineProperty(exports, "charAt", {
  enumerable: true,
  get: function get() {
    return _charAt.charAt;
  }
});
Object.defineProperty(exports, "concat", {
  enumerable: true,
  get: function get() {
    return _concat.concat;
  }
});
Object.defineProperty(exports, "create", {
  enumerable: true,
  get: function get() {
    return _create.create;
  }
});
Object.defineProperty(exports, "getActiveFormat", {
  enumerable: true,
  get: function get() {
    return _getActiveFormat.getActiveFormat;
  }
});
Object.defineProperty(exports, "getSelectionEnd", {
  enumerable: true,
  get: function get() {
    return _getSelectionEnd.getSelectionEnd;
  }
});
Object.defineProperty(exports, "getSelectionStart", {
  enumerable: true,
  get: function get() {
    return _getSelectionStart.getSelectionStart;
  }
});
Object.defineProperty(exports, "getTextContent", {
  enumerable: true,
  get: function get() {
    return _getTextContent.getTextContent;
  }
});
Object.defineProperty(exports, "isCollapsed", {
  enumerable: true,
  get: function get() {
    return _isCollapsed.isCollapsed;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function get() {
    return _isEmpty.isEmpty;
  }
});
Object.defineProperty(exports, "isEmptyLine", {
  enumerable: true,
  get: function get() {
    return _isEmpty.isEmptyLine;
  }
});
Object.defineProperty(exports, "join", {
  enumerable: true,
  get: function get() {
    return _join.join;
  }
});
Object.defineProperty(exports, "registerFormatType", {
  enumerable: true,
  get: function get() {
    return _registerFormatType.registerFormatType;
  }
});
Object.defineProperty(exports, "removeFormat", {
  enumerable: true,
  get: function get() {
    return _removeFormat.removeFormat;
  }
});
Object.defineProperty(exports, "remove", {
  enumerable: true,
  get: function get() {
    return _remove.remove;
  }
});
Object.defineProperty(exports, "replace", {
  enumerable: true,
  get: function get() {
    return _replace.replace;
  }
});
Object.defineProperty(exports, "insert", {
  enumerable: true,
  get: function get() {
    return _insert.insert;
  }
});
Object.defineProperty(exports, "insertLineSeparator", {
  enumerable: true,
  get: function get() {
    return _insertLineSeparator.insertLineSeparator;
  }
});
Object.defineProperty(exports, "insertObject", {
  enumerable: true,
  get: function get() {
    return _insertObject.insertObject;
  }
});
Object.defineProperty(exports, "slice", {
  enumerable: true,
  get: function get() {
    return _slice.slice;
  }
});
Object.defineProperty(exports, "split", {
  enumerable: true,
  get: function get() {
    return _split.split;
  }
});
Object.defineProperty(exports, "apply", {
  enumerable: true,
  get: function get() {
    return _toDom.apply;
  }
});
Object.defineProperty(exports, "unstableToDom", {
  enumerable: true,
  get: function get() {
    return _toDom.toDom;
  }
});
Object.defineProperty(exports, "toHTMLString", {
  enumerable: true,
  get: function get() {
    return _toHtmlString.toHTMLString;
  }
});
Object.defineProperty(exports, "toggleFormat", {
  enumerable: true,
  get: function get() {
    return _toggleFormat.toggleFormat;
  }
});
Object.defineProperty(exports, "LINE_SEPARATOR", {
  enumerable: true,
  get: function get() {
    return _specialCharacters.LINE_SEPARATOR;
  }
});
Object.defineProperty(exports, "unregisterFormatType", {
  enumerable: true,
  get: function get() {
    return _unregisterFormatType.unregisterFormatType;
  }
});

require("./store");

var _applyFormat = require("./apply-format");

var _charAt = require("./char-at");

var _concat = require("./concat");

var _create = require("./create");

var _getActiveFormat = require("./get-active-format");

var _getSelectionEnd = require("./get-selection-end");

var _getSelectionStart = require("./get-selection-start");

var _getTextContent = require("./get-text-content");

var _isCollapsed = require("./is-collapsed");

var _isEmpty = require("./is-empty");

var _join = require("./join");

var _registerFormatType = require("./register-format-type");

var _removeFormat = require("./remove-format");

var _remove = require("./remove");

var _replace = require("./replace");

var _insert = require("./insert");

var _insertLineSeparator = require("./insert-line-separator");

var _insertObject = require("./insert-object");

var _slice = require("./slice");

var _split = require("./split");

var _toDom = require("./to-dom");

var _toHtmlString = require("./to-html-string");

var _toggleFormat = require("./toggle-format");

var _specialCharacters = require("./special-characters");

var _unregisterFormatType = require("./unregister-format-type");
